src/garage/torch/algos/maml.py

```diff
@@ -0,0 +1,383 @@
+                [meta_batch_size * (num_grad_updates + 1)]
+            all_params (list[dict]): A list of named parameter dictionaries.
+                Each dictionary contains key value pair of names (str) and
+                parameters (torch.Tensor).
+
+        Returns:
+            float: Average return.
+
+        """
+        kl_before = self._compute_kl_constraint(itr,
+                                                all_samples,
+                                                all_params,
+                                                set_grad=False)
+
+        meta_objective = self._compute_meta_loss(itr, all_samples, all_params)
+
+        self._meta_optimizer.zero_grad()
+        meta_objective.backward()
+
+        self._meta_optimize(itr, all_samples, all_params)
+
+        # Log
+        loss_after = self._compute_meta_loss(itr,
+                                             all_samples,
+                                             all_params,
+                                             set_grad=False)
+        kl_after = self._compute_kl_constraint(itr,
+                                               all_samples,
+                                               all_params,
+                                               set_grad=False)
+
+        with torch.no_grad():
+            policy_entropy = self._compute_policy_entropy(
+                [task_samples[0] for task_samples in all_samples])
+            average_return = self._log(itr, all_samples, meta_objective.item(),
+                                       loss_after.item(), kl_before.item(),
+                                       kl_after.item(),
+                                       policy_entropy.mean().item())
+
+        return average_return
+
+    def _obtain_samples(self, runner):
+        """Obtain samples for each task before and after the fast-adaptation.
+
+        Args:
+            runner (LocalRunner): A local runner instance to obtain samples.
+
+        Returns:
+            tuple: Tuple of all_samples and all_params. all_samples is a
+                dimentional list of TorchTrajectoryBatch of size
+                [meta_batch_size * (num_grad_updates + 1)], and all_params is
+                a list of named parameter dictionaries.
+
+        """
+        tasks = self._env.sample_tasks(self._meta_batch_size)
+        all_samples = [[] for _ in range(len(tasks))]
+        all_params = []
+        theta = dict(self.policy.named_parameters())
+
+        for i, task in enumerate(tasks):
+            self._set_task(runner, task)
+
+            for j in range(self._num_grad_updates + 1):
+                paths = runner.obtain_samples(runner.step_itr)
+                batch_samples = self._process_samples(runner.step_itr, paths)
+                all_samples[i].append(batch_samples)
+
+                # Skip last iteration
+                if j != self._num_grad_updates:
+                    self._adapt(runner.step_itr, batch_samples, set_grad=False)
```

Comment:
Please correct me if I am wrong:
since `set_grad` is set to be `False`, these gradient is actually never applied to the policy during the sampling process? So in this case, each sampling step is using the same policy.